### ISSUE TITLE
Poor behavior around parameters that are both deprecated and required

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/sdk#48206]
Given the following code:
```dart
void f({@deprecated required int c}) {}
void g() {
  f(c: 2);
}
```
The analyzer marks the argument named `c` as being deprecated claiming that it shouldn't be used, but the call site is required to provide a value. There's no way for the user to fix the diagnostic, so we probably shouldn't produce it.

On the other hand, we should probably have a diagnostic indicating that it doesn't make sense to have a parameter that's both deprecated and required and suggest that users remove `required` and either make the parameter nullable or provide a default value.

**Comment by CicadaCinema** [comment]
@bwilkerson I had a go at adding a new diagnostic which fixes this: https://dart-review.googlesource.com/c/sdk/+/241008

Is there anything else I need to do here?

**Comment by bwilkerson** [comment]
Have you signed the Contributor License Agreement (https://cla.developers.google.com/)? I looked under 'CicadaCinema' but didn't find a signature. Maybe it's under a different name?

Just so you know, I was out for the first half of the week, so I have a lot of catching up to do, but will look at the CL as soon as possible.

**Comment by CicadaCinema** [comment]
> Have you signed the Contributor License Agreement (https://cla.developers.google.com/)? I looked under 'CicadaCinema' but didn't find a signature. Maybe it's under a different name?

I signed it about a year ago, but changed my GitHub username from 'ColonisationCaptain' since then - I did update it on the CLA page as well, so I really don't know why you can't see my signature:

![image](https://user-images.githubusercontent.com/52425971/163477642-2b39eb1b-1ac5-4e21-9c12-ebfdcbf9544d.png)

**Comment by bwilkerson** [comment]
Thanks. That should be sufficient.

I've added a couple of comments to the CL, but I have a more general question that I thought it would be better to discuss in an easier to find forum.

Should this also apply to required positional parameters? For example, should we report `p` in the following:
```
void f(@deprecated int p) {}
```
It seems like this has the same problem: invocation sites are required to provide a value but will then be told that the parameter is deprecated.

If so, we might want to update the correction message to suggest making the parameter optional (without specifying how) and adding a default value.

I think the current implementation already reports this case, but there's no test for it.

**Comment by CicadaCinema** [comment]
> invocation sites are required to provide a value but will then be told that the parameter is deprecated

I don't think this even happens for positional parameters:

```dart
void f(@deprecated int p) {}
void g() {
  f(1);
}
```
generates only one diagnostic telling us to provide a deprecation message:
```dart
$ dart analyze
Analyzing test_proj...                 1.4s

   info • lib/test_proj.dart:1:8 • Provide a deprecation message, via @Deprecated("message"). • provide_deprecation_message

1 issue found.
```

Perhaps we could move the check at https://github.com/dart-lang/sdk/blob/f398c9f3c1b190d50b4ca91f3b81a48830940616/pkg/analyzer/lib/src/error/deprecated_member_use_verifier.dart#L208 inside `_checkForDeprecated` (similar to [this](https://github.com/dart-lang/sdk/blob/f398c9f3c1b190d50b4ca91f3b81a48830940616/pkg/analyzer/lib/src/error/deprecated_member_use_verifier.dart#L137)) to fix the original issue, then deal with the new potential `deprecated_required_parameter` diagnosis separately at the declaration site?

**Comment by bwilkerson** [comment]
Yes, splitting up the CL is fine with me, and without looking at the code that seems like a reasonable approach.

**Comment by bwilkerson** [comment]
Given that Flutter has used deprecated required named parameters in the past (in `ThemeData`, and possibly in other places), and given that we use those deprecated diagnostics to be able to migrate user code forward, we can't remove the existing diagnostic (see https://dart-review.googlesource.com/c/sdk/+/241441). We might even need to add a similar diagnostic for deprecated required positional arguments at some point in order to move those API uses forward (though I don't really want to do so until we need to).

And given that it's been used, it seems likely that a check like this would have too many false positives to be a hint (enabled by default). We could consider implementing this as a lint (disabled by default), but the more I think about it, the less certain I am that even that would be a good idea, though I'm open to discussion.

While it certainly seems inconsistent from a user's perspective to be shown a diagnostic when they don't provide an argument (because the argument is required) and to be shown a diagnostic when they do provide an argument (because the parameter is deprecated), there really is no other alternative for required positional parameters. I could make an argument that named parameters should be made optional when they're deprecated, but it isn't always possible to do that for positional parameters. The only path forward for a required positional parameter (that doesn't require marking it as deprecated) is to mark the whole method as deprecated and create one with a new name.

**Comment by CicadaCinema** [comment]
> Given that Flutter has used deprecated required named parameters in the past

What is the reasoning behind this? Why would it be allowed behaviour?

**Comment by bwilkerson** [comment]
> What is the reasoning behind this? Why would it be allowed behaviour?

I'm not sure it's the case that someone thought "we should allow this"; it might well be the case that nobody thought "we should prohibit this". If it's the latter case, then once people start using the behavior it becomes more difficult to remove the ability to use it. In this particular case, removing the ability to deprecate required parameters might break our ability to migrate Flutter users new newer versions of Flutter, and that's likely too high a cost to pay. (If it was intentional, then I don't know what the reasoning was.)

@Piinks In case the discussion is interesting or you have insights to share.

**Comment by Piinks** [comment]
I agree that is does not make a lot of sense to require folks to use deprecated code.

In fact I think it is kinda contradictory. In the Flutter framework, we deprecate to allow people the opportunity to migrate their code comfortably. If they are being forced to use the deprecated API, they can't actually migrate off of it, and end up being broken later when we finally remove the API. I can audit the current deprecations in flutter/flutter and see if we can remove any cases like this. Filed for tracking https://github.com/flutter/flutter/issues/102106, I might be able to do this relatively quickly to unblock this, depends on where this may be happening. Thanks @bwilkerson!

**Comment by bwilkerson** [comment]
The one concrete example I know of is that the `dart fix` test in Flutter broke when we stopped reporting a diagnostic in cases where an argument's parameter was both required and deprecated. The concrete case(s) were in `ThemeData`. By not generating the diagnostic we lost the signal that enabled the fix.

The CL that stopped reporting this case was intended to be the first step toward reporting required and deprecated parameters at the declaration site rather than at the call site.

**Comment by Piinks** [comment]
Looks like it was an easy cleanup. I removed `required` where deprecated and all of the dart fixes still work as expected. :)

**Comment by bwilkerson** [comment]
Fantastic! Thanks.

**Comment by CicadaCinema** [comment]
Does this mean that after https://github.com/flutter/flutter/pull/102107 is merged we can re-merge https://github.com/dart-lang/sdk/commit/992618988befee18b6e964164ca94e074269f538 and keep working on the new declaration-site diagnostic ( https://dart-review.googlesource.com/c/sdk/+/241281 )?

**Comment by bwilkerson** [comment]
I think we can re-merge the reverted CL, yes. Whether we make the new diagnostic a hint or a lint might require a bit more discussion.

@srawlins @mit-mit For another case where `deprecated` has some interesting semantics.

**Comment by bwilkerson** [comment]
Given that there is no reasonable semantics, from a caller's point of view, for a parameter that is both required and deprecated, I'm inclined to enable this check by default.

@lrhn Your input would be welcome.

**Comment by lrhn** [comment]
There is no *actionable* input for a user of such a the method. If they call it, they must pass the required argument.
There is still signal in saying that the parameter goes away, but it's not really *useful* signal. All you can do is wait for the next major version release so you can stop passing and argument. You'd probably notice at that point too.

So I think we should issue a warning if a parameter is both deprecated and required, at the parameter declaration. The author can get rid of the warning by making the parameter optional, which provides users a migration path.
If the author chooses not to, I'd still show a deprecation warning at use places, because they did mark it `deprecated`.
